_posts/2019-08-29-using-tf-from-r/using-tf-from-R.Rmd

```diff
@@ -0,0 +1,413 @@
+
+```{Rcpp}
+operator PyObject*() const {
+  return get();
+}
+
+PyObject* get() const {
+  SEXP pyObject = getFromEnvironment("pyobj");
+  if (pyObject != R_NilValue) {
+    PyObject* obj = (PyObject*)R_ExternalPtrAddr(pyObject);
+    if (obj != NULL)
+      return obj;
+  }
+  Rcpp::stop("Unable to access object (object is from previous session and is now invalid)");
+}
+
+```
+
+So `py_to_r` works with a pointer to a Python object and returns what we want, an R object (a `SEXP`).
+The function checks for the type of the object, and then uses Rcpp to construct the adequate R object, in our case, an integer:
+
+```{Rcpp}
+
+else if (scalarType == INTSXP)
+  return IntegerVector::create(PyInt_AsLong(x));
+```
+
+For other objects, typically there's more action required; but essentially, the function is "just" a big `if`-`else` tree.
+
+So this was scenario 1: converting a Python object to R. Now in scenario 2, we assume we still need to create that Python object.
+
+### Scenario 2:
+
+As this scenario is considerably more complex than the previous one, we will explicitly concentrate on some aspects and leave out others. Importantly, we'll not go into module loading, which would deserve separate treatment of its own. Instead, we try to shed a light on what's involved using a concrete example: the ubiquitous, in `keras` code, `keras_model_sequential()`. All this R function does is
+  
+```{r}  
+function(layers = NULL, name = NULL) {
+  keras$models$Sequential(layers = layers, name = name)
+}
+```
+
+How can `keras$models$Sequential()` give us an object? When in Python, you run the equivalent
+
+```{python}
+tf.keras.models.Sequential()
+```
+
+this calls the constructor, that is, the `__init__` method of the class:
+
+```{python}
+class Sequential(training.Model):
+  def __init__(self, layers=None, name=None):
+    # ...
+  # ...
+```
+
+
+So this time, before -- as always, in the end -- getting an R object back from Python, we need to call that constructor, that is, a Python _callable_. (Python `callable`s subsume functions, constructors, and objects created from a class that has a `call` method.)
+
+So when `py_to_r`, inspecting it's argument's type, sees it is a Python callable (wrapped in a `PyObjectRef`, the `reticulate`-specific subclass of `Rcpp::Environment` we talked about above), it wraps it (the `PyObjectRef`) in an R function, using Rcpp:
+
+```{Rcpp}
+Rcpp::Function f = py_callable_as_function(pyFunc, convert);
+```
+
+The cpython-side action starts when `py_callable_as_function` then calls `py_call_impl`. `py_call_impl` executes the actual call and returns an R object, a `SEXP`. Now you may be asking, how does the Python runtime know it shouldn't deallocate that object, now that its work is done? This is taken of by the same `PyObjectRef` class used to wrap instances of `PyObject *`: It can wrap `SEXP`s as well.
```

Comment:
Oh, then I must have misunderstood - I thought this was because of the second constructor?

```
explicit PyObjectRef(SEXP object) : Rcpp::Environment(object) {}
```

Reply:
Thanks for explaining! (I think I won't put it in the post though, as that should be introductory.)